Batching; subkey auth; more namespace work

jagerman opened the pull request:
- adds "hf" tags to most responses (basically everywhere we have a "t" response).  Also fixes some places that were putting "t" inside the swarm member response instead of the outer request.
- disable namespace storing before HF19
- Delay retrieving requiring authentication until HF19.1
- Add subkey authentication for store/retrieve (this is draft support and may change, but if it doesn't this avoids needing a future SS update).
- Add batch/sequence requests; these are similar to SOGS in that "batch" does all the things, while "sequence" does them in order and aborts if anything in the sequence fails.
- Update/fix some documentation
- Change namespace signatures in various methods so that the namespace value in the signature is always "" for the default namespace; previously *some* namespace signatures required a "0" if the default namespace was explicitly given, "" if implicit, but this wasn't consistent and is potentially confusing.
- Fixed `expire_all` not having the namespace in the auth signature.

jagerman committed:
RPC: Add "hf" everywhere we current return "t"

jagerman committed:
Fix onion-request test script for refactored code

Also build it with cmake (but not by default), because figuring out the
necessary linking libs is a nuisance.

jagerman committed:
Don't allow storing to namespaces before HF19

jagerman committed:
Schedule delayed HF revision for requiring retrieve auth

jagerman committed:
Add draft subkey authentication for store/retrieve

This should be sufficient for the current draft design of v2 closed
groups; if so, including the support now saves us a soft fork later.

jagerman committed:
Refactor: move omq rpc code into rpc/request_handler

jagerman committed:
Add batch/sequence requests

jagerman committed:
Document RPC retrieve return value

jagerman committed:
Fix ts/hf being added nested rather than top-level

The initial SN contacted was adding the "t" and "hf" fields into its
swarm-specific result entry rather than into the top-level result.

jagerman committed:
Use "" instead of "0" for all default namespace signatures

This makes all endpoints consistent w.r.t. how the namespace is handled:
invoking with "namespace": 0 and omitting namespace are now equivalent
(and use the same signature value, with "" for the namespace value) for
all endpoints.

jagerman committed:
Add namespace to expire_all auth signature

jagerman committed:
Clarify timestamp/sig_timestamp use in store

The documentation used timestamp in some places that should have been
sig_timestamp (sig_timestamp default to timestamp, but they can be
different).

majestrate commented on oxenss/server/omq.cpp:
```diff
@@ -361,7 +321,7 @@ void OMQ::handle_client_request(std::string_view method, oxenmq::Message& messag
     try {
         std::string_view params = message.data.size() == full_size ? message.data.back() : ""sv;
-        it->second(
+        it->second.omq(
```
will this properly throw if `it == rpc::RequestHandler::client_rpc_endpoints.end()` ?

jagerman replied:
It shouldn't be possible because the code calling this is responsible for only passing in a valid method; there's a C-assert earlier in the function to check it in debug builds, but the check shouldn't be needed.

majestrate commented on oxenss/rpc/request_handler.cpp:
```diff
@@ -1034,6 +1055,84 @@ void RequestHandler::process_client_req(rpc::expire_msgs&& req, std::function<vo
+        rpc::sequence&& req, std::function<void(rpc::Response)> cb) {
+
+    if (req.subreqs.empty()) {
+        cb(Response{http::OK, json({{"results", json::array()}})});
```
should this be a 400? it seems like a client should not be able to send an empty array and have it be okay with it.

jagerman replied:
I'm kind of indifferent about it; I'll change it.

jagerman committed:
Disallow empty batch/sequence; fix docs

Change "up to 6" -> "1 to 5" in the docs; the latter was a typo, the
former for this commit.

jagerman committed:
Comment hairiness in batch/sequence handling

jagerman committed:
Fix bug in namespace retrieval

The retrieval queries were only half-using the namespace id and
returning things from all namespaces.  Fixed.

jagerman committed:
Fix drone format script

jagerman committed:
Add `ifelse` branching metarequest

This lets you make conditional requests based on the HF, SS version, or
block height.  For future use (since this won't be usable before HF19).

jagerman committed:
Allow extending expiry (up to TTL from now)

This removes the restriction on updating message expiries that only
allowed shortening, but not extending, expiries.

With this change, a client can renew its own messages indefinitely as
long as it continually re-renews within 14 days of the last renewal.

majestrate commented on oxenss/rpc/client_rpc_endpoints.cpp:
```diff
@@ -728,15 +753,147 @@ void batch::load_from(bt_dict_consumer params) {
+    if (!(height_ge_ || height_lt_ || hf_ge || hf_lt || v_ge || v_lt))
+        throw parse_error{"Invalid ifelse request: must specify at least one \"if\" condition"};
+
+    i.condition = [=](const snode::ServiceNode& snode) {
```
this will capture `i`and `if_` by value, probably making a copy, very expensive and it does not use them in that lambda

jagerman replied:
`[=]` will only copy variables that you use in the lambda

majestrate commented on oxenss/rpc/client_rpc_endpoints.h:
```diff
@@ -603,8 +615,117 @@ struct sequence : batch {
+///       }
+///     }
+///
+struct ifelse : endpoint {
```
aw yea now all we need is a jump instruction and we will have our own turning complete json format.

majestrate commented on oxenss/rpc/request_handler.cpp:
```diff
@@ -102,14 +102,12 @@ namespace {
+        calls.load_req_json = [](json params) -> client_request {
+            return load_request<RPC>(std::move(params));
+        };
+        calls.load_req_bt = [](oxenc::bt_dict_consumer params) -> client_request {
```
this feels like these two should be one function that takes a variant in.

jagerman replied:
Good idea, done.

majestrate commented on oxenss/storage/database.cpp:
```diff
@@ -776,27 +776,24 @@ std::vector<std::string> Database::update_expiry(
     st.bind(1, new_exp_ms);
-    st.bind(2, new_exp_ms);
-    bind_pubkey(st, 3, 4, pubkey);
+    bind_pubkey(st, 2, 3, pubkey);
```
i feel like there could be an iterator approach for binding parameters, in which you create an iterator and it is incremented as you bind, by each call, as needed.

jagerman replied:
There is (sort of) for most cases (e.g. binding a fixed set of values), but here it made a bit more sense to me to see where the 1,2,3,4+i numbers are coming from.

majestrate replied:
it feels like a footcannon potential.
